File: implementations/java/SLD.java
package io.github.proteo5.sld;

import java.util.*;

/**
 * SLD/MLD (Single/Multi Line Data) Format - Java Implementation v2.0
 * A token-efficient data serialization format
 *
 * Changes in v2.0:
 * - Field separator changed from | to ; (semicolon)
 * - Added MLD format support (records separated by newlines)
 * - Array marker changed to { (curly brace)
 * - Property marker remains [ (square bracket)
 */
public class SLDParser {
    // Constants
    private static final char FIELD_SEPARATOR = ';';
    private static final char RECORD_SEPARATOR_SLD = '~';
    private static final char RECORD_SEPARATOR_MLD = '\n';
    private static final char PROPERTY_MARKER = '[';
    private static final char ARRAY_MARKER = '{';
    private static final char ESCAPE_CHAR = '^';

    /**
     * Escape special SLD/MLD characters
     */
    public static String escapeValue(String text) {
        if (text == null || text.isEmpty()) {
            return "";
        }

        return text
            .replace("^", "^^")
            .replace(";", "^;")
            .replace("~", "^~")
            .replace("[", "^[")
            .replace("{", "^{");
    }

    /**
     * Unescape SLD/MLD escape sequences
     */
    public static Object unescapeValue(String text) {
        StringBuilder result = new StringBuilder();
        int i = 0;

        while (i < text.length()) {
            if (text.charAt(i) == ESCAPE_CHAR && i + 1 < text.length()) {
                char nextChar = text.charAt(i + 1);
                if (nextChar == '1') {
                    return true; // Boolean true
                } else if (nextChar == '0') {
                    return false; // Boolean false
                } else {
                    result.append(nextChar);
                }
                i += 2;
            } else {
                result.append(text.charAt(i));
                i++;
            }
        }

        return result.toString();
    }

    /**
     * Split text by delimiter, respecting escape sequences
     */
    private static List<String> splitUnescaped(String text, char delimiter) {
        List<String> parts = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        int i = 0;

        while (i < text.length()) {
            if (text.charAt(i) == ESCAPE_CHAR && i + 1 < text.length()) {
                current.append(text.substring(i, i + 2));
                i += 2;
            } else if (text.charAt(i) == delimiter) {
                parts.add(current.toString());
                current = new StringBuilder();
                i++;
            } else {
                current.append(text.charAt(i));
                i++;
            }
        }

        if (current.length() > 0 || text.endsWith(String.valueOf(delimiter))) {
            parts.add(current.toString());
        }

        return parts;
    }

    /**
     * Encode record to string
     */
    private static String encodeRecord(Map<String, Object> record) {
        List<String> parts = new ArrayList<>();

        for (Map.Entry<String, Object> entry : record.entrySet()) {
            String escapedKey = escapeValue(entry.getKey());
            Object value = entry.getValue();

            if (value instanceof Map) {
                // Nested object
                @SuppressWarnings("unchecked")
                String nested = encodeRecord((Map<String, Object>) value);
                parts.add(escapedKey + PROPERTY_MARKER + nested);
            } else if (value instanceof List) {
                // Array
                @SuppressWarnings("unchecked")
                List<Object> list = (List<Object>) value;
                List<String> nestedItems = new ArrayList<>();
                for (Object item : list) {
                    nestedItems.add(escapeValue(item.toString()));
                }
                parts.add(escapedKey + ARRAY_MARKER + String.join(",", nestedItems));
            } else if (value instanceof Boolean) {
                // Boolean
                String boolVal = (Boolean) value ? "^1" : "^0";
                parts.add(escapedKey + PROPERTY_MARKER + boolVal);
            } else if (value == null) {
                // Null as ^_
                parts.add(escapedKey + PROPERTY_MARKER + "^_");
            } else {
                // Regular value
                String escapedValue = escapeValue(value.toString());
                parts.add(escapedKey + PROPERTY_MARKER + escapedValue);
            }
        }

        return String.join(String.valueOf(FIELD_SEPARATOR), parts);
    }

    /**
     * Encode data to SLD format
     */
    public static String encodeSLD(List<Map<String, Object>> data) {
        List<String> records = new ArrayList<>();
        for (Map<String, Object> record : data) {
            records.add(encodeRecord(record));
        }
        return String.join(String.valueOf(RECORD_SEPARATOR_SLD), records) + RECORD_SEPARATOR_SLD;
    }

    /**
     * Encode single record to SLD format
     */
    public static String encodeSLD(Map<String, Object> data) {
        return encodeRecord(data);
    }

    /**
     * Encode data to MLD format
     */
    public static String encodeMLD(List<Map<String, Object>> data) {
        List<String> records = new ArrayList<>();
        for (Map<String, Object> record : data) {
            records.add(encodeRecord(record));
        }
        return String.join(String.valueOf(RECORD_SEPARATOR_MLD), records);
    }

    /**
     * Encode single record to MLD format
     */
    public static String encodeMLD(Map<String, Object> data) {
        return encodeRecord(data);
    }

    /**
     * Decode record string
     */
    private static Map<String, Object> decodeRecord(String recordStr) {
        Map<String, Object> record = new LinkedHashMap<>();
        List<String> fields = splitUnescaped(recordStr, FIELD_SEPARATOR);

        for (String field : fields) {
            if (field.isEmpty()) {
                continue;
            }

            // Check for property marker
            if (field.contains(String.valueOf(PROPERTY_MARKER)) && 
                !field.contains(ESCAPE_CHAR + String.valueOf(PROPERTY_MARKER))) {
                String[] parts = field.split("\\\\" + PROPERTY_MARKER, 2);
                String key = unescapeValue(parts[0]).toString();
                String value = parts.length > 1 ? unescapeValue(parts[1]).toString() : "";
                
                // Handle null (^_)
                if ("^_".equals(value)) {
                    record.put(key, null);
                } else {
                    record.put(key, value);
                }
            }
            // Check for array marker
            else if (field.contains(String.valueOf(ARRAY_MARKER)) && 
                     !field.contains(ESCAPE_CHAR + String.valueOf(ARRAY_MARKER))) {
                String[] parts = field.split("\\" + ARRAY_MARKER, 2);
                String key = unescapeValue(parts[0]).toString();
                if (parts.length > 1) {
                    List<Object> items = new ArrayList<>();
                    for (String item : parts[1].split(",")) {
                        items.add(unescapeValue(item));
                    }
                    record.put(key, items);
                } else {
                    record.put(key, new ArrayList<>());
                }
            }
        }

        return record;
    }

    /**
     * Decode SLD format string
     */
    public static Object decodeSLD(String sldString) {
        if (sldString == null || sldString.isEmpty()) {
            return new LinkedHashMap<String, Object>();
        }

        sldString = sldString.replaceAll("~+$", "");
        List<Map<String, Object>> records = new ArrayList<>();

        for (String recordStr : splitUnescaped(sldString, RECORD_SEPARATOR_SLD)) {
            if (!recordStr.isEmpty()) {
                records.add(decodeRecord(recordStr));
            }
        }

        return records.size() > 1 ? records : (records.size() == 1 ? records.get(0) : new LinkedHashMap<String, Object>());
    }

    /**
     * Decode MLD format string
     */
    public static Object decodeMLD(String mldString) {
        if (mldString == null || mldString.isEmpty()) {
            return new LinkedHashMap<String, Object>();
        }

        List<Map<String, Object>> records = new ArrayList<>();
        String[] lines = mldString.split("\n");

        for (String line : lines) {
            String trimmed = line.trim();
            if (!trimmed.isEmpty()) {
                records.add(decodeRecord(trimmed));
            }
        }

        return records.size() > 1 ? records : (records.size() == 1 ? records.get(0) : new LinkedHashMap<String, Object>());
    }

    /**
     * Convert SLD to MLD
     */
    public static String sldToMLD(String sldString) {
        return sldString.replaceAll("~+$", "").replace('~', '\n');
    }

    /**
     * Convert MLD to SLD
     */
    public static String mldToSLD(String mldString) {
        return mldString.replace('\n', '~') + '~';
    }

    // Example usage
    public static void main(String[] args) {
        System.out.println("=== SLD/MLD Java Implementation v2.0 ===\n");

        // Example 1: Simple records with SLD
        System.out.println("Example 1: Simple user data (SLD)");
        List<Map<String, Object>> data1 = new ArrayList<>();
        Map<String, Object> user1 = new LinkedHashMap<>();
        user1.put("name", "Alice");
        user1.put("age", 30);
        user1.put("city", "New York");
        Map<String, Object> user2 = new LinkedHashMap<>();
        user2.put("name", "Bob");
        user2.put("age", 25);
        user2.put("city", "Los Angeles");
        data1.add(user1);
        data1.add(user2);

        String sld1 = encodeSLD(data1);
        System.out.println("Encoded SLD: " + sld1);
        System.out.println("Decoded: " + decodeSLD(sld1) + "\n");

        // Example 2: Booleans
        System.out.println("Example 2: Boolean values");
        Map<String, Object> data2 = new LinkedHashMap<>();
        data2.put("name", "Alice");
        data2.put("verified", true);
        data2.put("active", false);

        String sld2 = encodeSLD(data2);
        System.out.println("Encoded SLD: " + sld2);
        System.out.println("Decoded: " + decodeSLD(sld2) + "\n");

        // Example 3: Format conversion
        System.out.println("Example 3: Format conversion");
        String sld3 = "name[Alice;age[30~name[Bob;age[25~";
        String mld3 = sldToMLD(sld3);
        System.out.println("SLD: " + sld3);
        System.out.println("MLD:\n" + mld3);
        System.out.println("Back to SLD: " + mldToSLD(mld3));
    }
}

File: implementations/csharp/SLD.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SLD
{
    /// <summary>
    /// SLD/MLD (Single/Multi Line Data) Format - C# Implementation v1.1
    /// A token-efficient data serialization format
    /// 
    /// Changes in v1.1:
    /// - Field separator changed from | to ; (semicolon)
    /// - Added MLD format support (records separated by newlines)
    /// - Array marker changed to { (curly brace)
    /// - Property marker remains [ (square bracket)
    /// </summary>
    public static class SLDParser
    {
        // Constants
        private const char FIELD_SEPARATOR = ';';
        private const char RECORD_SEPARATOR_SLD = '~';
        private const char RECORD_SEPARATOR_MLD = '\n';
        private const char PROPERTY_MARKER = '[';
        private const char ARRAY_MARKER = '{';
        private const char ESCAPE_CHAR = '^';

        /// <summary>
        /// Escape special SLD/MLD characters in a string
        /// </summary>
        public static string EscapeValue(string text)
        {
            if (string.IsNullOrEmpty(text))
                return string.Empty;

            return text
                .Replace("^", "^^")
                .Replace(";", "^;")
                .Replace("~", "^~")
                .Replace("[", "^[")
                .Replace("{", "^{");
        }

        /// <summary>
        /// Unescape SLD/MLD escape sequences
        /// </summary>
        public static object UnescapeValue(string text)
        {
            var result = new StringBuilder();
            int i = 0;

            while (i < text.Length)
            {
                if (text[i] == ESCAPE_CHAR && i + 1 < text.Length)
                {
                    char nextChar = text[i + 1];
                    if (nextChar == '1')
                        return true; // Boolean true
                    else if (nextChar == '0')
                        return false; // Boolean false
                    else
                        result.Append(nextChar);
                    i += 2;
                }
                else
                {
                    result.Append(text[i]);
                    i++;
                }
            }

            return result.ToString();
        }

        /// <summary>
        /// Encode data to SLD format
        /// </summary>
        public static string EncodeSLD(List<Dictionary<string, object>> data)
        {
            var records = data.Select(record => EncodeRecord(record));
            return string.Join(RECORD_SEPARATOR_SLD.ToString(), records) + RECORD_SEPARATOR_SLD;
        }

        /// <summary>
        /// Encode single dictionary to SLD format
        /// </summary>
        public static string EncodeSLD(Dictionary<string, object> data)
        {
            return EncodeRecord(data);
        }

        /// <summary>
        /// Encode data to MLD format
        /// </summary>
        public static string EncodeMLD(List<Dictionary<string, object>> data)
        {
            var records = data.Select(record => EncodeRecord(record));
            return string.Join(RECORD_SEPARATOR_MLD.ToString(), records);
        }

        /// <summary>
        /// Encode single dictionary to MLD format
        /// </summary>
        public static string EncodeMLD(Dictionary<string, object> data)
        {
            return EncodeRecord(data);
        }

        private static string EncodeRecord(Dictionary<string, object> record)
        {
            var parts = new List<string>();

            foreach (var kvp in record)
            {
                string escapedKey = EscapeValue(kvp.Key);
                object value = kvp.Value;

                if (value is Dictionary<string, object> dict)
                {
                    // Nested object
                    string nested = EncodeRecord(dict);
                    parts.Add($"{escapedKey}{PROPERTY_MARKER}{nested}");
                }
                else if (value is List<object> list)
                {
                    // Array using { marker
                    var nestedItems = list.Select(item => EscapeValue(item.ToString()));
                    parts.Add($"{escapedKey}{ARRAY_MARKER}{string.Join(",", nestedItems)}");
                }
                else if (value is bool boolValue)
                {
                    // Boolean as ^1 or ^0
                    string boolVal = boolValue ? "^1" : "^0";
                    parts.Add($"{escapedKey}{PROPERTY_MARKER}{boolVal}");
                }
                else if (value == null)
                {
                    // Null value
                    parts.Add($"{escapedKey}{PROPERTY_MARKER}");
                }
                else
                {
                    // Regular value
                    string escapedValue = EscapeValue(value.ToString());
                    parts.Add($"{escapedKey}{PROPERTY_MARKER}{escapedValue}");
                }
            }

            return string.Join(FIELD_SEPARATOR.ToString(), parts);
        }

        /// <summary>
        /// Decode SLD format string
        /// </summary>
        public static object DecodeSLD(string sldString)
        {
            if (string.IsNullOrEmpty(sldString))
                return new Dictionary<string, object>();

            sldString = sldString.TrimEnd(RECORD_SEPARATOR_SLD);

            var records = new List<Dictionary<string, object>>();
            var recordStrings = SplitUnescaped(sldString, RECORD_SEPARATOR_SLD);

            foreach (var recordStr in recordStrings)
            {
                if (!string.IsNullOrEmpty(recordStr))
                    records.Add(DecodeRecord(recordStr));
            }

            return records.Count > 1 ? (object)records : (records.Count == 1 ? records[0] : new Dictionary<string, object>());
        }

        /// <summary>
        /// Decode MLD format string
        /// </summary>
        public static object DecodeMLD(string mldString)
        {
            if (string.IsNullOrEmpty(mldString))
                return new Dictionary<string, object>();

            var records = new List<Dictionary<string, object>>();
            var lines = mldString.Split(new[] { RECORD_SEPARATOR_MLD }, StringSplitOptions.RemoveEmptyEntries);

            foreach (var line in lines)
            {
                if (!string.IsNullOrWhiteSpace(line))
                    records.Add(DecodeRecord(line));
            }

            return records.Count > 1 ? (object)records : (records.Count == 1 ? records[0] : new Dictionary<string, object>());
        }

        private static Dictionary<string, object> DecodeRecord(string recordStr)
        {
            var record = new Dictionary<string, object>();
            var fields = SplitUnescaped(recordStr, FIELD_SEPARATOR);

            foreach (var field in fields)
            {
                if (string.IsNullOrEmpty(field))
                    continue;

                // Check for property marker
                if (field.Contains(PROPERTY_MARKER) && !field.Contains($"{ESCAPE_CHAR}{PROPERTY_MARKER}"))
                {
                    var parts = field.Split(new[] { PROPERTY_MARKER }, 2);
                    string key = UnescapeValue(parts[0]).ToString();
                    object value = parts.Length > 1 ? UnescapeValue(parts[1]) : null;
                    record[key] = value;
                }
                // Check for array marker
                else if (field.Contains(ARRAY_MARKER) && !field.Contains($"{ESCAPE_CHAR}{ARRAY_MARKER}"))
                {
                    var parts = field.Split(new[] { ARRAY_MARKER }, 2);
                    string key = UnescapeValue(parts[0]).ToString();
                    if (parts.Length > 1)
                    {
                        var items = parts[1].Split(',').Select(item => UnescapeValue(item)).ToList();
                        record[key] = items;
                    }
                    else
                    {
                        record[key] = new List<object>();
                    }
                }
            }

            return record;
        }

        private static List<string> SplitUnescaped(string text, char delimiter)
        {
            var parts = new List<string>();
            var current = new StringBuilder();
            int i = 0;

            while (i < text.Length)
            {
                if (text[i] == ESCAPE_CHAR && i + 1 < text.Length)
                {
                    current.Append(text[i]);
                    current.Append(text[i + 1]);
                    i += 2;
                }
                else if (text[i] == delimiter)
                {
                    parts.Add(current.ToString());
                    current.Clear();
                    i++;
                }
                else
                {
                    current.Append(text[i]);
                    i++;
                }
            }

            if (current.Length > 0 || text.EndsWith(delimiter.ToString()))
            {
                parts.Add(current.ToString());
            }

            return parts;
        }

        /// <summary>
        /// Convert SLD to MLD format
        /// </summary>
        public static string SLDToMLD(string sldString)
        {
            return sldString.TrimEnd(RECORD_SEPARATOR_SLD).Replace(RECORD_SEPARATOR_SLD, RECORD_SEPARATOR_MLD);
        }

        /// <summary>
        /// Convert MLD to SLD format
        /// </summary>
        public static string MLDToSLD(string mldString)
        {
            return mldString.Replace(RECORD_SEPARATOR_MLD, RECORD_SEPARATOR_SLD) + RECORD_SEPARATOR_SLD;
        }
    }

    // Example usage class
    public class Program
    {
        public static void Main(string[] args)
        {
            Console.WriteLine("=== SLD/MLD C# Implementation v1.1 ===\n");

            // Example 1: Simple records with SLD
            Console.WriteLine("Example 1: Simple user data (SLD)");
            var data1 = new List<Dictionary<string, object>>
            {
                new Dictionary<string, object> { {"name", "Alice"}, {"age", 30}, {"city", "New York"} },
                new Dictionary<string, object> { {"name", "Bob"}, {"age", 25}, {"city", "Los Angeles"} }
            };
            string sld1 = SLDParser.EncodeSLD(data1);
            Console.WriteLine($"Encoded SLD: {sld1}");
            Console.WriteLine($"Decoded: {SLDParser.DecodeSLD(sld1)}\n");

            // Example 2: Booleans
            Console.WriteLine("Example 2: Boolean values");
            var data2 = new Dictionary<string, object>
            {
                {"name", "Alice"},
                {"verified", true},
                {"active", false}
            };
            string sld2 = SLDParser.EncodeSLD(data2);
            Console.WriteLine($"Encoded SLD: {sld2}");
            Console.WriteLine($"Decoded: {SLDParser.DecodeSLD(sld2)}\n");

            // Example 3: Format conversion
            Console.WriteLine("Example 3: Format conversion");
            string sld3 = "name[Alice;age[30~name[Bob;age[25~";
            string mld3 = SLDParser.SLDToMLD(sld3);
            Console.WriteLine($"SLD: {sld3}");
            Console.WriteLine($"MLD:\n{mld3}");
            Console.WriteLine($"Back to SLD: {SLDParser.MLDToSLD(mld3)}");
        }
    }
}
